mqtt/mqtt-bridge/src/persist/memory/waking_map.rs

```diff
@@ -0,0 +1,229 @@
+use std::{cmp::min, collections::HashMap, collections::VecDeque, task::Waker};
+
+use mqtt3::proto::Publication;
+use tracing::error;
+
+use crate::persist::Key;
+
+/// Responsible for waking waiting streams when new elements are added
+/// Exposes a get method for retrieving a count of elements starting from queue head
+/// Once elements are retrieved they are added to the in flight collection
+pub struct WakingMap {
+    queue: VecDeque<(Key, Publication)>,
+    in_flight: HashMap<Key, Publication>,
```

Comment:
The in-flight messages for the in-memory persistence implementation don't have a clear purpose.

For the disk case (not this PR) it makes sense. For disk case, messages cannot be removed from the main collection right away because in case of crash we want them to be resent. Thus a separate in-flight collection is needed to prevent duplicates from being returned from the loader. When an element is removed form the in-flight it can be removed from the main persistence.

For the in-mem case everything is lost in the case of a crash anyway, so the in-flight collection doesn't have the same purpose. Removing from the queue immediately on retrieval by the loader allows for a simple queue to be used. The messages can be removed from the main queue directly when retrieved by the loader, rather than linger around in the main queue until being removed from in-flight. I think having this in-flight collection is still necessary for logging, but we need to find an appropriate place to do that.

If we cannot think of a place to log in-flight, then we can remove the in-flight collection from this in-memory implementation.

Reply:
I now disagree with this. In-flight serves the purpose of only allowing elements to be removed from the collection which have been extracted by the loader component.